tesseract_core/runtime/finite_differences.py

```diff
@@ -0,0 +1,551 @@
+                    exc_info = f"{type(e).__name__}: '{e}' in file {tb[-1].filename}, line {tb[-1].lineno}"
+                    failure = GradientCheckResult(
+                        in_path=in_path,
+                        out_path=out_path,
+                        idx=idx,
+                        ref_val=None,
+                        grad_val=None,
+                        exception=exc_info,
+                    )
+                else:
+                    if not np.allclose(result_apply, result_grad, atol=1e-8, rtol=rtol):
+                        failure = GradientCheckResult(
+                            in_path=in_path,
+                            out_path=out_path,
+                            idx=idx,
+                            ref_val=result_apply,
+                            grad_val=result_grad,
+                            exception=None,
+                        )
+
+                if failure:
+                    failures.append(failure)
+                    progress.update(
+                        subtask,
+                        description=f"Checking gradients for {endpoint}... (failures: {len(failures)})",
+                    )
+
+                progress.update(subtask, advance=1)
+
+    except BaseException as e:
+        # Somtimes, Pydantic re-raises exceptions as Pydantic<...>Exception so we check the string representation
+        is_interrupt = isinstance(e, KeyboardInterrupt) or "KeyboardInterrupt" in str(e)
+        if not is_interrupt:
+            raise
+        print("Interrupted")
+
+    return failures, num_evals
+
+
+def check_gradients(
+    api_module: ModuleType,
+    inputs: dict[str, Any],
+    *,
+    input_paths: Optional[Sequence[str]] = None,
+    output_paths: Optional[Sequence[str]] = None,
+    base_dir: Path,
+    endpoints: Optional[Sequence[ADEndpointName]] = None,
+    max_evals: int = 1000,
+    eps: float = 1e-4,
+    rtol: float = 0.1,
+    seed: Optional[int] = None,
+    show_progress: bool = True,
+):
+    """Check gradients of endpoints against a finite difference approximation.
+
+    Args:
+        api_module: The module containing the Tesseract endpoints.
+        inputs: The inputs to apply to evaluate gradients at.
+        input_paths: The input paths to check. If not provided, all differentiable paths are checked.
+        output_paths: The output paths to check. If not provided, all differentiable paths are checked.
+        base_dir: The base directory to resolve relative paths.
+        endpoints: The AD endpoints to check. If not provided, all available endpoints are checked.
+        max_evals: The target number of ``apply`` evaluations to perform.
+        eps: The epsilon to use for finite differences, as a fraction of the maximum absolute value of each input.
+        rtol: The relative tolerance to use for comparison.
+        seed: The random seed to use for sampling. If not provided, a random seed is used.
```

Comment:
```suggestion
        seed: The random seed to use for sampling. If not provided, no random seed is used.
```

Reply:
I think that is actually correct: if you do not provide a seed, a random one will be picked

Reply:
ah, right. so "random random seed"?